src/REAXFF/fix_qtpie_reaxff.cpp

```diff
@@ -0,0 +1,1231 @@
+  }
+
+  b_norm = parallel_norm(b, nn);
+  sig_new = parallel_dot(r, d, nn);
+
+  for (i = 1; i < imax && sqrt(sig_new) / b_norm > tolerance; ++i) {
+    comm->forward_comm(this); //Dist_vector(d);
+    sparse_matvec(&H, d, q);
+    comm->reverse_comm(this); //Coll_vector(q);
+
+    tmp = parallel_dot(d, q, nn);
+    alpha = sig_new / tmp;
+
+    vector_add(x, alpha, d, nn);
+    vector_add(r, -alpha, q, nn);
+
+    // pre-conditioning
+    for (jj = 0; jj < nn; ++jj) {
+      j = ilist[jj];
+      if (atom->mask[j] & groupbit)
+        p[j] = r[j] * Hdia_inv[j];
+    }
+
+    sig_old = sig_new;
+    sig_new = parallel_dot(r, p, nn);
+
+    beta = sig_new / sig_old;
+    vector_sum(d, 1., p, beta, d, nn);
+  }
+
+  if ((i >= imax) && maxwarn && (comm->me == 0))
+    error->warning(FLERR, "Fix qtpie/reaxff CG convergence failed after {} iterations at step {}",
+                   i,update->ntimestep);
+  return i;
+}
+
+
+/* ---------------------------------------------------------------------- */
+
+void FixQtpieReaxFF::sparse_matvec(sparse_matrix *A, double *x, double *b)
+{
+  int i, j, itr_j;
+  int ii;
+
+  for (ii = 0; ii < nn; ++ii) {
+    i = ilist[ii];
+    if (atom->mask[i] & groupbit)
+      b[i] = eta[atom->type[i]] * x[i];
+  }
+
+  int nall = atom->nlocal + atom->nghost;
+  for (i = atom->nlocal; i < nall; ++i)
+      b[i] = 0;
+
+  for (ii = 0; ii < nn; ++ii) {
+    i = ilist[ii];
+    if (atom->mask[i] & groupbit) {
+      for (itr_j=A->firstnbr[i]; itr_j<A->firstnbr[i]+A->numnbrs[i]; itr_j++) {
+        j = A->jlist[itr_j];
+        b[i] += A->val[itr_j] * x[j];
+        b[j] += A->val[itr_j] * x[i];
+      }
+    }
+  }
+
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixQtpieReaxFF::calculate_Q()
+{
+  int i, k;
+  double u, s_sum, t_sum;
+  double *q = atom->q;
+
+  int ii;
+
+  s_sum = parallel_vector_acc(s, nn);
+  t_sum = parallel_vector_acc(t, nn);
+  u = s_sum / t_sum;
+
+  for (ii = 0; ii < nn; ++ii) {
+    i = ilist[ii];
+    if (atom->mask[i] & groupbit) {
+      q[i] = s[i] - u * t[i];
+
+      /* backup s & t */
+      for (k = nprev-1; k > 0; --k) {
+        s_hist[i][k] = s_hist[i][k-1];
+        t_hist[i][k] = t_hist[i][k-1];
+      }
+      s_hist[i][0] = s[i];
+      t_hist[i][0] = t[i];
+    }
+  }
+
+  pack_flag = 4;
+  comm->forward_comm(this); //Dist_vector(atom->q);
+}
+
+/* ---------------------------------------------------------------------- */
+
+int FixQtpieReaxFF::pack_forward_comm(int n, int *list, double *buf,
+                                  int /*pbc_flag*/, int * /*pbc*/)
+{
+  int m;
+
+  if (pack_flag == 1)
+    for (m = 0; m < n; m++) buf[m] = d[list[m]];
+  else if (pack_flag == 2)
+    for (m = 0; m < n; m++) buf[m] = s[list[m]];
+  else if (pack_flag == 3)
+    for (m = 0; m < n; m++) buf[m] = t[list[m]];
+  else if (pack_flag == 4)
+    for (m = 0; m < n; m++) buf[m] = atom->q[list[m]];
+  else if (pack_flag == 5) {
```

Comment:
Is this option ever used? If so, the *comm_forward = 1* assignment in the constructor is insufficient. Ditto for reverse comm.

Reply:
This is a good point. The `pack_flag = 5` option is never used (it is only used in `fix qeq/reaxff/omp`) and has now been removed.